core/unit_test/TestViewCtorProp.hpp

```diff
@@ -0,0 +1,106 @@
+using vcp_label_base_t = Kokkos::Impl::ViewCtorProp<void, std::string>;
+using vcp_label_t      = Kokkos::Impl::ViewCtorProp<std::string>;
+
+// Check traits of Kokkos::Impl::ViewCtorProp<>.
+TEST(TEST_CATEGORY, vcp_empty_traits) {
+  // Check that the empty view constructor properties class is default
+  // constructible. This is needed for calls of Kokkos::view_alloc().
+  static_assert(std::is_default_constructible_v<vcp_empty_t>);
+  static_assert(std::is_same_v<decltype(Kokkos::view_alloc()), vcp_empty_t>);
+}
+
+// Check Kokkos::Impl::is_view_label.
+TEST(TEST_CATEGORY, is_view_label) {
+  static_assert(Kokkos::Impl::is_view_label<std::string>::value);
+
+  constexpr unsigned N = 3;
+  static_assert(Kokkos::Impl::is_view_label<const char[N]>::value);
+  static_assert(Kokkos::Impl::is_view_label<char[N]>::value);
+
+  // A char* is not a label. Thus, a label is distinguished from a pointer type.
+  static_assert(!Kokkos::Impl::is_view_label<char*>::value);
+}
+
+// Check traits of base class Kokkos::Impl::ViewCtorProp<void, std::string>.
+TEST(TEST_CATEGORY, vcp_label_base_traits) {
+  static_assert(std::is_same_v<typename vcp_label_base_t::type, std::string>);
+
+  // Check that the base class is default constructible. The default constructor
+  // may be called by the copy constructor of deriveded classes, such as when
+  // copy constructing a view constructor properties object from another view
+  // constructor properties object that holds fewer properties.
+  static_assert(std::is_default_constructible_v<vcp_label_base_t>);
+}
+
+// Check traits of derived class Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_traits) {
+  static_assert(std::is_base_of_v<vcp_label_base_t, vcp_label_t>);
+
+  static_assert(vcp_label_t::has_label);
+
+  // Check that the derived class is not default constructible. It is a design
+  // choice to not allow the default constructor to be called.
+  static_assert(!std::is_default_constructible_v<vcp_label_t>);
+}
+
+// Check that Kokkos::view_alloc perfect forwards a label passed by
+// rvalue reference, and check that the constructor
+// of Kokkos::Impl::ViewCtorProp<std::string> moves this label.
+TEST(TEST_CATEGORY, view_alloc_can_perfect_forward_label) {
+  std::string label("our label");
+
+  vcp_label_t prop = Kokkos::view_alloc(std::move(label));
+
+  ASSERT_TRUE(label.empty());
+  ASSERT_EQ(Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop),
+            "our label");
+}
+
+// Check the copy constructor of Kokkos::Impl::ViewCtorProp<std::string>.
+TEST(TEST_CATEGORY, vcp_label_copy_constructor) {
+  // Copy construction from an empty view constructor properties object.
+  static_assert(std::is_constructible_v<vcp_label_t, const vcp_empty_t&>);
+
+  vcp_empty_t prop_empty;
+  vcp_label_t prop_empty_copy(prop_empty);
+
+  ASSERT_TRUE(
+      Kokkos::Impl::get_property<Kokkos::Impl::LabelTag>(prop_empty_copy)
+          .empty());
```

Comment:
This behavior is currently needed because of how `Impl::with_properties_if_unset` is currently implemented:

- https://github.com/kokkos/kokkos/blob/22fea8395ac617d3b5a933e8cc2905a4c3b3b6f4/core/src/View/Kokkos_ViewCtor.hpp#L252-L270

There are current use cases where `view_ctor_prop` on line 267 is empty, and `new_view_ctor_prop` on this line is not empty, e.g. it has a label, or an execution space instance, or ... Thus, with this test, we want to make sure that such current use cases work.

I agree that this behavior is a bit strange. It may make sense to refactor `Impl::with_properties_if_unset` to avoid it. The test could then be deleted. May be best in a follow-up PR? I would be happy to work on it.

It's fine either way for me. I can also delete the test now if you prefer.

Reply:
As I understand it, his PR is about properly moving label strings into view alloc property construct and it is orthogonal to what `Impl::with_properties_if_unset()` does, meaning that test addition does not really belong here.

Reply:
OK. No problem. I have just deleted this test that constructs a copy from an empty view alloc prop. Perhaps when we have a chance, we can discuss live whether or not we'd like to do a follow-up pr. It could be a follow-up pr that adds move semantics to `Impl::with_properties_if_unset()` and also addresses the issue of how it relies on the strange behavior of the copy constructor of the view alloc prop.